gtsfm/densify/mvsnets/methods/PatchmatchNet/eval_gtsfm.py

```diff
@@ -0,0 +1,435 @@
+"""Customized MVSNet evaluation functions for gtsfm
+    reference: https://github.com/FangjinhuaWang/PatchmatchNet
+
+Authors: Ren Liu
+"""
+
+import os
+import time
+from typing import Dict, List, Tuple
+
+import cv2
+import numpy as np
+from PIL import Image
+from plyfile import PlyData, PlyElement
+import torch
+import torch.nn as nn
+import torch.nn.parallel
+import torch.backends.cudnn as cudnn
+from torch.utils.data import DataLoader
+
+from gtsfm.densify.mvsnets.methods.PatchmatchNet.datasets.gtsfm import MVSDataset
+from gtsfm.densify.mvsnets.methods.PatchmatchNet.models.net import PatchmatchNet
+from gtsfm.densify.mvsnets.methods.PatchmatchNet.utils import tocuda, tensor2numpy
+from gtsfm.densify.mvsnets.methods.PatchmatchNet.datasets.data_io import save_pfm
+import gtsfm.utils.logger as logger_utils
+
+logger = logger_utils.get_logger()
+
+cudnn.benchmark = True
+
+
+def read_img(img: np.ndarray, img_wh: Tuple[int, int]) -> np.ndarray:
+    """read image and resize
+    Args:
+        img: input image
+        img_wh: width and height of target size
+    Returns:
+        np.array image
+    """
+    # scale 0~255 to 0~1
+    np_img = np.array(img, dtype=np.float32) / 255.0
+    np_img = cv2.resize(np_img, img_wh, interpolation=cv2.INTER_LINEAR)
+    return np_img
+
+
+def save_mask(filename: str, mask: np.ndarray) -> None:
+    """save a binary mask
+    Args:
+        filename: output file string
+        mask: output mask
+    """
+    assert mask.dtype == np.bool
+    mask = mask.astype(np.uint8) * 255
+    Image.fromarray(mask).save(filename)
+
+
+def save_depth_img(filename: str, depth: np.ndarray) -> None:
+    """save a depth map as image
+    Args:
+        filename: output file string
+        depth: output depth map
+    """
+    d = depth.max() - depth.min()
+    b = -depth.min()
+    depth = 255 * (depth + b) / d
+    transform = np.array([d / 255.0, -b])
+    depth = depth.astype(np.uint8)
+
+    cv2.imwrite(filename, depth)
+    np.save(filename + ".npy", transform)
+
+
+def parse_pairs(pairs: np.ndarray, nviews: int) -> List[Tuple[int, List[int]]]:
+    """prepare candidate pairs for each view from pairwise distances
+    Args:
+        pairs: pre-calculated pairwise distances
+        nviews: number of views, or candidate views as pairs
+    Returns:
+        For each tuple, is each view's id and list of pair views' indices
+    """
+    num_viewpoint = pairs.shape[0]
+
+    for i in range(num_viewpoint):
+        pairs[i][i] = -np.inf
+
+    pair_idx = np.argsort(pairs, axis=0)[:, 1:nviews][:, ::-1]
+
+    data = []
+    for view_idx in range(num_viewpoint):
+        ref_view = view_idx
+        src_views = pair_idx[i].tolist()
+        data.append((ref_view, src_views))
+    return data
+
+
+def reproject_with_depth(
+    depth_ref: np.ndarray,
+    intrinsics_ref: np.ndarray,
+    extrinsics_ref: np.ndarray,
```

Comment:
what is the GTSAM equivalent? Likely it is `i1Tw` and `i2Tw`? We should discuss this in person